### PR TITLE
node-init: Fixing pods restart on nodes running containerd on COS

### LINKED ISSUE #13167: CiliumEndpoint IP does not match Pod IP
## Bug report

There is a CiliumEndpoint IP and Pod IP mismatch for one of the `kube-dns` Pods. This results in dropped traffic because Cilium thinks the target IP is unmanaged and thus not affected by network policies allowing traffic.

```
kubectl -n kube-system get po kube-dns-7c976ddbdb-77p6w -owide
NAME                        READY   STATUS    RESTARTS   AGE   IP          NODE                        NOMINATED NODE   READINESS GATES
kube-dns-7c976ddbdb-77p6w   4/4     Running   0          5d    10.12.3.7   gke-dev-dev-39261880-f5rb   <none>           <none>
```

```
kubectl -n kube-system get cep kube-dns-7c976ddbdb-77p6w
NAME                        ENDPOINT ID   IDENTITY ID   INGRESS ENFORCEMENT   EGRESS ENFORCEMENT   ENDPOINT STATE   IPV4         IPV6
kube-dns-7c976ddbdb-77p6w   1645          16773543                                                 ready            10.12.3.93
```

Have only seen this happen for one of the `kube-dns` Pods (although the phenomenon is present in multiple clusters), but it may affect other Pods without us noticing yet.

**General Information**

- Cilium version (run `cilium version`)
Client: 1.8.1 5ce2bc7b3 2020-07-02T20:04:47+02:00 go version go1.14.4 linux/amd64
Daemon: 1.8.1 5ce2bc7b3 2020-07-02T20:04:47+02:00 go version go1.14.4 linux/amd64
- Kernel version (run `uname -a`)
Linux gke-dev-dev-39261880-f5rb 4.19.112+ #1 SMP Thu May 21 12:32:38 PDT 2020 x86_64 x86_64 x86_64 GNU/Linux
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.13-gke.1", GitCommit:"688c6543aa4b285355723f100302d80431e411cc", GitTreeState:"clean", BuildDate:"2020-07-21T02:37:26Z", GoVersion:"go1.13.9b4", Compiler:"gc", Platform:"linux/amd64"}
- Link to relevant artifacts (policies, deployments scripts, ...)
- Generate and upload a system zip:

**How to reproduce the issue**

1. Create a GKE cluster
2. Install Cilium (following https://docs.cilium.io/en/stable/gettingstarted/k8s-install-gke/). Cilium is installed in the `cilium` namespace. Only difference being `nodeinit.restartPods=true` and `global.kubeProxyReplacement=disabled`
3. Check the `kube-dns` Pod IP:s and their respective CiliumEndpoint IP:s.
`kubectl -n kube-system get po -l k8s-app=kube-dns -owide`
`kubectl -n kube-system get cep | grep kube-dns`

### PR DESCRIPTION
**Pull request by fallard84**
The startup script for the node-init was not properly working for nodes running containerd on COS for two reasons:
 1. crictl is not available in the PATH for root
 2. extracting the sandbox id using cat was returning an extra line

This fix is necessary for deployment on GKE using preemptible nodes. Otherwise, pods on nodes getting preempted might be created before Cilium is ready to manage them and they will remain that way unless they are properly stopped.

Fixes: #13167

```release-note
Fixing pods restart on nodes running containerd on COS
```

### CODE REVIEW

File: install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
Comment (aanm) [inline install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml]: How does it help running "which crictl"? Is it to make the script existing since it couldn't find such command and exit the node-init?

Comment (fallard84) [reply install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml]: I just found it useful debugging information. I can remove if you don't believe it is relevant to keep.

File: install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
Comment (ti-mo) [inline install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml]: I would call `crictl` by its absolute path instead of manipulating `PATH`. Why is `crictl` in `cni.binPath`? :thinking: Has this changed in recent versions of COS?

**Review by ti-mo — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
Comment (fallard84) [reply install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml]: I am not sure if that changed in previous version of COS. But I agree that it would be better to explicitely call `crictl` through its absolute path: `/home/kubernetes/bin/crictl`. I'll change that.

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR. Change LGTM, Can you squash the second commit into the first?

File: install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
Comment (fallard84) [reply install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml]: Removed it as I am now calling crictl using the absolute path.

**Review by christarazi — APPROVED** [review]

**Review by ti-mo — APPROVED** [review]
Looks good, thanks for the fix!

**Review by kaworu — APPROVED** [review]
LGTM, thanks for the PR!

NOTE: no need for a full CI run.

**Review by aanm — CHANGES_REQUESTED** [review]
@ti-mo @fallard84 we can't merge the PR as is at the risk of breaking older COS images. Can we either use `crictl` from PATH, and if not available, fallback to `/home/kubernetes/bin/crictl`? Thanks

**Comment by fallard84** [comment]
> @ti-mo @fallard84 we can't merge the PR as is at the risk of breaking older COS images. Can we either use `crictl` from PATH, and if not available, fallback to `/home/kubernetes/bin/crictl`? Thanks

Yeah, we could do that. I'll make the change. Do you know which COS version had crictl in the PATH for root?

**Comment by aanm** [comment]
> > @ti-mo @fallard84 we can't merge the PR as is at the risk of breaking older COS images. Can we either use `crictl` from PATH, and if not available, fallback to `/home/kubernetes/bin/crictl`? Thanks
> 
> Yeah, we could do that. I'll make the change. Do you know which COS version had crictl in the PATH for root?

No, sorry. What I can guarantee is at the time the script was created the `crictl` was in the PATH :smile:

**Comment by fallard84** [comment]
I encountered another issue related to this (discussed in [Slack](https://cilium.slack.com/archives/C1MATJ5U5/p1612221599275900) with @tgraf ). Basically there is a race condition issue where pods get created after the node-init stopped the pods but before the cilium-agent is ready to manage the network. So we end up with pods that are running and are not managed by Cilium, breaking a lot of the inter-pods policies.

I thought about opening a separate PR and close this one, since this will be replaced anyway. Let me know if you prefer to go forward with this small fix and merge the other fix afterward. I should open the other PR today.

File: install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
Comment (ti-mo) [inline install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml]: - FYI: `which` is not a shell built-in (at least not in Bash), the convention is to use `command -v` (see https://stackoverflow.com/questions/592620/how-can-i-check-if-a-program-exists-from-a-bash-script for explanation).
- Checking whether or not `crictl` exists here is arguably a bit pointless. There's no harm in having `/home/kubernetes/bin` in PATH even on older COS installs where this dir is not present.
- Please add a comment stating why this PATH extension is done. (e.g. required as of COS version x.y.z or any other useful info you might have) This helps with maintenance.

Thanks!

**Review by ti-mo — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
Comment (fallard84) [reply install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml]: Thanks, I wasn't aware that `which` wasn't recommended. I will make the requested changes.

**Review by joestringer — APPROVED** [review]
Thanks for the submission! No docs-structure impact, approving from that codeowner. The other feedback will still need to be addressed before merge though.

**Comment by joestringer** [comment]
@fallard84 are you still planning to follow up on the feedback on this PR? I'll mark this PR as a draft for now, then when you would like additional review, scroll to the bottom of the page and click the "Ready for review" button and we'll take another look.

**Comment by fallard84** [comment]
> @fallard84 are you still planning to follow up on the feedback on this PR? I'll mark this PR as a draft for now, then when you would like additional review, scroll to the bottom of the page and click the "Ready for review" button and we'll take another look.

Sorry for the delay. I will do this in the next few days.

**Comment by joestringer** [comment]
No worries, let us know how we can help. :slightly_smiling_face:

**Comment by aanm** [comment]
@fallard84 any update? We are about to release 1.10 and we would like to have this included :-)

**Comment by aanm** [comment]
This was tested locally. Merging

**Pull request merged**
